lightning/src/ln/features.rs

```diff
@@ -395,6 +399,64 @@ impl InvoiceFeatures {
 	}
 }
 
+impl ToBase32 for InvoiceFeatures {
+	fn write_base32<W: WriteBase32>(&self, writer: &mut W) -> Result<(), <W as WriteBase32>::Err> {
+		let total_u5s = ((self.flags.len() as f64 * 8 as f64) / 5 as f64).ceil() as usize;
+		let mut be_u5s: Vec<u5> = vec![u5::try_from_u8(0).unwrap(); total_u5s];
```

Comment:
Note I was able to get this to work with the bech32 native stuff, but its not as succinct as I'd like, take it or leave it:

```
let mut vec_read = bech32::convert_bits(field_data, 5, 8, true)?;
let mut good_bytes: Vec<u8>;
let bitshift = (field_data.len() * 5) % 8;

if bitshift == 0 {
	// If we don't need to shift, just byte-swap vec_read
	for i in 0..vec_read.len() / 2 {
		let tmp = vec_read[i];
		let swap_idx = vec_read.len() - i - 1;
		vec_read[i] = vec_read[swap_idx];
		vec_read[swap_idx] = tmp;
	}
	good_bytes = vec_read;
} else {
	// If we didn't read the right bit count, we need to drop bitshift bits at the end while byte swapping
	good_bytes = vec![0; total_bytes];
	for i in 0..total_bytes {
		let read_pos = vec_read.len() - i - 1;
		if i == total_bytes - 1 {
			good_bytes[i] = vec_read[read_pos] >> (8-bitshift);
		} else {
			good_bytes[i] = (vec_read[read_pos] >> (8-bitshift)) | (vec_read[read_pos - 1] << bitshift);
		}
	}
}
while !good_bytes.is_empty() && good_bytes[good_bytes.len() - 1] == 0 {
	good_bytes.pop();
}
Ok(InvoiceFeatures::from_le_bytes(good_bytes))
```

Reply:
Ah, cool. Probably going to leave it but lmk if there's a strong reason to take it

Reply:
No, up to you, I was just annoyed that you said it wasn't working so had to fix it :).

Reply:
Actually, I take that back, reading each bit one by one is definitely not the way to do this, especially if we branch on each bit. the above code works, but we can definitely write our own reader too, just as long as we aren't doing it bit-by-bit.

Reply:
Ok, I guess in my mind there's not that many feature bits. Is this specifically for performance or also for a DoS vector reasons? Fine to switch to your impl^ above

Reply:
Right, my proposal above sucks, but there's not really a reason to do per-bit iteration, this works fine on the read end


```
		let mut res_bytes: Vec<u8> = vec![0; total_bytes];
		for (u5_idx, chunk) in field_data.iter().enumerate() {
			let bit_pos_from_right_0_indexed = (field_data.len() - u5_idx - 1) * 5;
			let new_byte_idx = (bit_pos_from_right_0_indexed / 8) as usize;
			let new_bit_pos = bit_pos_from_right_0_indexed % 8;
			let chunk_u16 = chunk.to_u8() as u16;
			res_bytes[new_byte_idx] |= ((chunk_u16 << new_bit_pos) & 0xff) as u8;
			if new_byte_idx != total_bytes - 1 {
				res_bytes[new_byte_idx + 1] |= ((chunk_u16 >> (8-new_bit_pos)) & 0xff) as u8;
			}
		}
```